### ISSUE TITLE
Location Sensor Location Changed Reports 0,0

### ISSUE THREAD

**Issue by ram8647** [mit-cml/appinventor-sources#606]
This is an issue that's been reported by at least a dozen people in the edX course.  Apparently, on some devices, the LocationSensor.LocationChanged block reports the lat,long as 0,0.   I haven't noticed this happening on any of my devices but it shouldn't be possible.  

Looking at the Texting source code, latitude and longitude are initialized to **UNKNOWN_VALUE**, which is defined as 0.   The **onLocationChanged()** method (below) sets longitude and latitude without checking for UNKNOWN_VALUE. Apparently, the Android *onLocationChanged()* method can be called before the device has gotten a fix on its latitude and longitude.   Android documentation is not clear about this. 

And note that *onLocationChanged()* (below) also sets **hasLocationData** to true.  This variable is used elsewhere in the code to set the  **HasLongitudeLatitude** property to true.  If lat,long is 0,0, the sensor should not be reporting that it HasLongitudeLatitude and should not be firing the LocationChanged block.    

```javascript
public void onLocationChanged(Location location) {
      lastLocation = location;
      longitude = location.getLongitude();
      latitude = location.getLatitude();
      // If the current location doesn't have altitude information, the prior
      // altitude reading is retained.
      if (location.hasAltitude()) {
        hasAltitude = true;
        altitude = location.getAltitude();
      }
      hasLocationData = true;
      LocationChanged(latitude, longitude, altitude);
    }
```
I think this bug can be fixed by adding the following lines to *onLocationChanged*:

```javascript
      if (longitude != UNKNOWN_VALUE && latitude != UNKNOWN_VALUE) {
         hasLocationData = true;
         LocationChanged(latitude, longitude, altitude);
      }
```
In other words, **LocationChanged** will not be fired when lat,long = 0,0.  If we want to allow 0,0 as a legitimate location -- it is a valid lat,long -- then we should use some some value other than 0 as UNKNOWN_VALUE (e.g., something outside the range 0..360).

**Comment by ram8647** [comment]
Doing some additional research on this issue.  

Apparently, by default, [Location.latitude and Location.longitude are set to 0,0](http://developer.android.com/reference/android/location/Location.html#Location(java.lang.String)) when a Location object is created.   

And according to this thread on [Stackoverflow](http://stackoverflow.com/questions/32650872/latitude-and-longitude-values-are-giving-null-values-0-0),  

>There are various reasons to get a null or (0.0, 0.) coordinate. Altough there exists a valid (0.0, 0.0) >coordinate, in practise it is never a valid one, or designed that way to express invalid locations.
>So you can safely ignore all null and (0.0, 0.0) locations.

So, I propose leaving UNKNOWN_VALUE as 0 and merely checking for it in the onLocationChanged() method, as described above.

**Comment by halatmit** [comment]
pr #646 was closed

**Comment by JaimeChavarriaga** [comment]
Nowadays, applications using the Location Sensor does not perform well. They become unresponsive because the position may remain as 0,0 for a long time. 

In my current class, we may spent a lot of time and the location component never gets a new position. As a workaround, we enable and disable the location sensor (manually) in the phone keeping the app inventor application active. If we start another activity (e.g. Google Maps), the other app can get the current position, but when we return to the application the position remains as 0,0. 
  - This happens at least with the phones in the lab: Samsung Galaxy Alpha, Samsung Galaxy A5 2016, Samsung Galaxy A5 2017 and Galaxy Note 10.1 devices; and with all the phones of my students, including some ASUS and Sony Xperia phones.

Google Maps is able to trigger a "get current location" when it starts. I suppose App Inventor can make the same when the application starts or when the screen gets active if it includes a location component.
